module/zfs/vdev_removal.c

```diff
@@ -1995,19 +1995,44 @@ spa_vdev_remove_top_check(vdev_t *vd)
 		if (available < vd->vdev_stat.vs_alloc)
 			return (SET_ERROR(ENOSPC));
 	} else {
-		/* available space in the pool's normal class */
-		uint64_t available = dsl_dir_space_available(
-		    spa->spa_dsl_pool->dp_root_dir, NULL, 0, B_TRUE);
-		if (available <
-		    vd->vdev_stat.vs_dspace + spa_get_slop_space(spa)) {
-			/*
-			 * This is a normal device. There has to be enough free
-			 * space to remove the device and leave double the
-			 * "slop" space (i.e. we must leave at least 3% of the
-			 * pool free, in addition to the normal slop space).
-			 */
+		vdev_stat_t *rs = &spa->spa_root_vdev->vdev_stat;
+		vdev_stat_t *vs = &vd->vdev_stat;
+
+		/*
+		 * This is a normal device. There has to be enough free
+		 * space in the pool's normal class to copy the allocated
+		 * space from the device being removed, this will account
+		 * for both reserved and "slop" space.  Any available
+		 * capacity on the device being removed is deducted from
+		 * the total available pool capacity.
+		 */
+		int64_t available = dsl_dir_space_available(
+		    spa->spa_dsl_pool->dp_root_dir, NULL, 0, B_TRUE) -
+		    (vs->vs_dspace - vs->vs_alloc);
+
+		/*
+		 * Calculate the "slop" space requirement based on the pool
+		 * capacity after the device has been removed.  This helps
+		 * avoid the case where removal is prevented because a large
+		 * vdev is added to a small pool resulting in a vastly
+		 * inflated slop requirement.  Furthermore, if the "slop"
+		 * space exceeds a quarter of the total pool capacity then
+		 * the pool must be tiny and this requirement is reduced.
+		 */
+		int64_t slop = spa_get_slop_space(spa, -vs->vs_dspace);
+		int64_t factor = 4;
+
+		if (slop > (rs->vs_dspace - vs->vs_dspace) / 4)
+			factor = 1;
+
+		/*
+		 * There has to be enough free space to remove the device
+		 * and leave double the "slop" space (i.e. we must leave
+		 * at least 3% of the pool free, in addition to the normal
+		 * slop space).
+		 */
+		if (available < (int64_t)vs->vs_alloc + (factor * slop / 2))
```

Comment:
I'm not sure that the requirement to leave some AVAIL space is really related to the slop size, and it's resulting in adding more complication to the rule here.  It might be simpler to change this rule to something like "2% of the pool must remain AVAIL after the removal", regardless of pool size.  Also then we won't have to revisit this when changing the calculation of slop space (e.g. in https://github.com/openzfs/zfs/pull/11389).

Reply:
I've updated the PR to drop the additional slop requirement and instead require that ~3% of the pool remain AVAIL after the removal.  This works well in my testing and is by far simpler.